### PR TITLE
OffsetPath and ZigZag modifiers

### PR DESCRIPTION
**Pull request by mbasaglia**
Implements support OffsetPath (`op`) and ZigZag (`zz`) modifiers.

See https://github.com/airbnb/lottie-web/pull/2810 for more context, this merge request builds on that one but adds offset path and a new bezier utility module with all the operations I need for it. I kept it separate from `utils/bez.js`.

For some reason OffsetPath lags a lot when dev tools are open but it works fine with that closed.

### CODE REVIEW

File: player/js/utils/shapes/ZigZagModifier.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  ShapeModifier,
+} from './ShapeModifiers';
+import { PolynomialBezier } from '../PolynomialBezier';
+
+function ZigZagModifier() {}
+extendPrototype([ShapeModifier], ZigZagModifier);
+ZigZagModifier.prototype.initModifierProperties = function (elem, data) {
+  this.getValue = this.processKeys;
+  this.amplitude = PropertyFactory.getProp(elem, data.s, 0, null, this);
+  this.frequency = PropertyFactory.getProp(elem, data.pt, 0, null, this);
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:16]: `data.pt` refers to the type of points the zigzag uses. I think here `data.r` is needed which refers to the Ridges per segment, or frequency as it's called here.

File: player/js/utils/shapes/ZigZagModifier.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  if (!path.c) {
+    pathLength -= 1;
+  }
+
+  for (var cur = 0; cur < pathLength; cur += 1) {
+    var bez = PolynomialBezier.shapeSegment(path, cur);
+
+    vX = bez.points[0][0];
+    vY = bez.points[0][1];
+    clonedPath.setTripleAt(vX, vY, vX, vY, vX, vY, clonedPath.length());
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:37]: Comparing the behavior of AE with the result of this calculation, it seems that this point should also be offset by the amplitude.  
I'm looking into how to the tangent is calculated for it.

File: player/js/utils/shapes/ZigZagModifier.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    for (var i = 0; i < frequency; i += 1) {
+      var t = (i + 0.5) / frequency;
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:40]: `var t = (i + 1) / (frequency + 1);`
This  will split the path into the right amount of segments

File: player/js/utils/shapes/ZigZagModifier.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+ZigZagModifier.prototype.processShapes = function (_isFirstFrame) {
+  var shapePaths;
+  var i;
+  var len = this.shapes.length;
+  var j;
+  var jLen;
+  var amplitude = this.amplitude.v;
+  var frequency = Math.max(1, Math.round(this.frequency.v));
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:66]: No need to max it at 1, it will also work with 0

File: player/js/utils/shapes/ZigZagModifier.js (hunk 2 above)
Comment (bodymovin) [reply player/js/utils/shapes/ZigZagModifier.js:37]: After some more tests, I figured out that the vertices need to be offset by a vector tangent to its surrounding vertices.

File: player/js/utils/shapes/ZigZagModifier.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  this._isAnimated = this.amplitude.effectsSequence.length !== 0 && this.frequency.effectsSequence.length !== 0;
+};
+
+ZigZagModifier.prototype.processPath = function (path, amplitude, frequency) {
+  var pathLength = path._length;
+  var clonedPath = shapePool.newElement();
+  clonedPath.c = path.c;
+  var vX;
+  var vY;
+  var direction = 1;
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:26]: direction should start at `-1` and be flipped immedately after offsetting the first corner

File: player/js/utils/shapes/ZigZagModifier.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      var pt = bez.point(t);
+      var normal = bez.normalAngle(t);
+      vX = pt[0] + Math.cos(normal) * direction * amplitude;
+      vY = pt[1] - Math.sin(normal) * direction * amplitude;
+
+      clonedPath.setTripleAt(vX, vY, vX, vY, vX, vY, clonedPath.length());
+      direction = -direction;
+    }
+
+    direction = -direction;
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:50]: this should me moved after line 37

File: player/js/utils/shapes/OffsetPathModifier.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
+import {
```
Comment (bodymovin) [inline player/js/utils/shapes/OffsetPathModifier.js:1]: This is amazing! I've ran some tests and many of them work great.
There are some cases that break, for example with triangles, or some combinations with other shape modifiers, but it's definitely a great first version. I'd feel comfortable pushing this one as it is and keep improving it when there's feedback.

File: player/js/utils/shapes/ZigZagModifier.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  for (var cur = 0; cur < pathLength; cur += 1) {
+    var bez = PolynomialBezier.shapeSegment(path, cur);
+
+    vX = bez.points[0][0];
+    vY = bez.points[0][1];
+    clonedPath.setTripleAt(vX, vY, vX, vY, vX, vY, clonedPath.length());
+
+    for (var i = 0; i < frequency; i += 1) {
+      var t = (i + 0.5) / frequency;
+      var pt = bez.point(t);
```
Comment (bodymovin) [inline player/js/utils/shapes/ZigZagModifier.js:41]: For bezier curves with control points that don't match the vertices, this pt doesn't seem to be calculated right. It's most noticeable with curves with pronounced control point values.
I haven't done enough tests, but I'm pretty sure this is related to how AE splits the path in equal length segments instead of getting the "t" values at equidistant positions of the [0, 1] range.  
I've had to fix this in the past for trim path, and the solution is splitting the curve into multiple segments (in the order of hundreds) and find the right "t" for the corresponding value.  
This will take some time, and performance will degrade a little since calculating it is not that straightforward.  
I'm ok merging it the way it is and iterating over it on a following PR.

File: player/js/utils/shapes/ZigZagModifier.js (hunk 2 above)
Comment (mbasaglia) [reply player/js/utils/shapes/ZigZagModifier.js:37]: Offset by the bezier tangent?

File: player/js/utils/shapes/OffsetPathModifier.js (hunk 7 above)
Comment (mbasaglia) [reply player/js/utils/shapes/OffsetPathModifier.js:1]: yeah if the path is offset inward too much on sharp corners and the joints overlap (rather than the offset segments) it breaks. My first thought was to check for intersections between line joints and when there is one to remove the segments between them but for some reason I haven't been able to implement it properly

File: player/js/utils/shapes/ZigZagModifier.js (hunk 8 above)
Comment (mbasaglia) [reply player/js/utils/shapes/ZigZagModifier.js:41]: Perhaps a way to optimize it could be to just split the bezier in a handful of segments and assume it's linear within those segments. This will introduce small errors in length calculations but it shouldn't be too noticeable (Based on some quick tests if you split each segments in 10 linear chunks you get ~2% error)

**Comment by mbasaglia** [comment]
Latest commit made a lot of changes to zigzag to handle the offsets on the segment end points, it also avoids duplicating them (unlike the previous version). I _think_ it handles all possible corner cases with angles and such

**Comment by bodymovin** [comment]
All my tests passed. I'll try to work on the zigzag curved paths as soon as I can.
Thanks! This is great.
